subcommands/targets/offline-update.go

```diff
@@ -0,0 +1,264 @@
+		Run:   doOfflineUpdate,
+		Args:  cobra.ExactArgs(2),
+	}
+	cmd.AddCommand(offlineUpdateCmd)
+	offlineUpdateCmd.Flags().BoolVarP(&prod, "prod", "", false,
+		"Instruct to fetch content of production Target")
+	offlineUpdateCmd.Flags().StringVarP(&tag, "tag", "", "master",
+		"Target tag")
+}
+
+func doOfflineUpdate(cmd *cobra.Command, args []string) {
+	factory = viper.GetString("factory")
+	targetName := args[0]
+	dstDir := args[1]
+
+	fmt.Printf("Downloading offline update content of Target %s to %s\n", targetName, dstDir)
+
+	fmt.Printf("Getting info about Target %s \n", targetName)
+
+	// We get the Target's custom data from the list of all targets regardless whether it's a production target or not
+	// because we need to obtain the Target's tag that was used to build OE and Apps/containers, which in turn is needed
+	// to deduce a path to the Apps archive artifact.
+	// E.g. https://ci.foundries.io/projects/msul-dev01/lmp/builds/1420/assemble-system-image/ is built as `master`
+	// while in the production targets the corresponding target is marked as `prod`.
+	// TODO: switch to usage of /api/v1/user_repo/targets.json API once it supports `production` param
+	// instead of doing the stuff below
+	custom, err := getTargetCustomInfo(targetName)
+	subcommands.DieNotNil(err)
+
+	if !prod && custom.Tags[0] != tag {
+		fmt.Fprintf(os.Stderr, "The specified Target hasn't been found; target: %s, tag: %s\n",
+			targetName, tag)
+		os.Exit(1)
+	}
+
+	if prod {
+		isProdTarget, err := checkIfProdTarget(targetName)
+		subcommands.DieNotNil(err)
+		if !isProdTarget {
+			fmt.Fprintf(os.Stderr, "The specified Target is not a production target; target %s, tag: %s\n",
+				targetName, tag)
+			os.Exit(1)
+		}
+	}
+
+	targetVer, err := strconv.Atoi(custom.Version)
+	subcommands.DieNotNil(err)
+
+	fmt.Println("Downloading TUF metadata...")
+	err = downloadTufRepo(factory, tag, prod, path.Join(dstDir, "tuf"))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to download TUF repo: %s", err.Error())
+		os.Exit(1)
+	}
+
+	hardwareID := custom.HardwareIds[0]
+	buildTag := custom.Tags[0]
+	targetOstreeVer := targetVer
+	if len(custom.OrigUri) > 0 {
+		indx := strings.LastIndexByte(custom.OrigUri, '/')
+		if indx == -1 {
+			fmt.Fprintf(os.Stderr, "Failed to determine Target's OE build version: %s", custom.OrigUri)
+			os.Exit(1)
+		}
+		targetOstreeVerStr := custom.OrigUri[indx+1:]
+		targetOstreeVer, err = strconv.Atoi(targetOstreeVerStr)
+		subcommands.DieNotNil(err)
+	}
+	fmt.Printf("Downloading an ostree repo from the Target's OE build %d...\n", targetOstreeVer)
+	err = downloadOstree(targetOstreeVer, hardwareID, dstDir)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to download Target ostree repo: %s", err.Error())
+		os.Exit(1)
+	}
+
+	fmt.Printf("Downloading Apps fetched by the `assemble-system-image` run of the build:  %d...\n", targetVer)
+	err = downloadApps(targetName, targetVer, buildTag, path.Join(dstDir, "apps"))
```

Comment:
IIRC most of our updates are either ostree-only or apps-only i.e. an ostree update versus an apps update are normally delivered as two separate targets to the devices (via online updates). I think that offline updates should support that behavior as well (i.e. allow ostree-only or apps-only updates). AFAIK in case of waves and production targets this is the only way how it works i.e. a user delivers one or more ostree-only updates followed by one or more apps-only updates for each tag.

I don't know how it is planned to implement the offline update on the device side, but IMHO supporting ostree-only and apps-only updates is at least a necessity, and at most may be even the only supported way to do this.

Reply:
It implies knowing exactly which version is currently running on each target device, and those versions may differ. How to guarantee that the Apps of a given offline update can be supported by all ostree versions devices currently running and vice versa.

It's not big deal to support it, I'd postpone with this feature customization once we get some feedback about initial implementation. Maybe we'll get zero customers for it :).